File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import           Data.Char (isLetter, isSpace, toUpper)

responseFor :: String -> String
responseFor = let
  respond t
    | isSilence t = "Fine. Be that way!"
    | isShouting t && isQuestion t = "Calm down, I know what I'm doing!"
    | isShouting t = "Whoa, chill out!"
    | isQuestion t = "Sure."
    | otherwise = "Whatever."
  in respond . strip

isShouting :: String -> Bool
isShouting s = any isLetter s && all (\c -> c == toUpper c) s

isQuestion :: String -> Bool
isQuestion = (== '?') . last

isSilence :: String -> Bool
isSilence = (== 0) . length

strip :: String -> String
strip = filter (not . isSpace)

File: clojure/bob/src/bob.clj
(ns bob
  (:require [clojure.string]))

(defn asking? [input]
  (= \? (last input)))

(defn yelling? [input]
  (and (some #(Character/isLetter %) input)
       (= input (clojure.string/upper-case input))))

(defn saying-nothing? [input]
  (= "" input))

(defn response-for [s]
  (let [input (clojure.string/trim s)]
    (cond (saying-nothing? input) "Fine. Be that way!"
          (and (asking? input) (yelling? input))
          "Calm down, I know what I'm doing!"
          (asking? input) "Sure."
          (yelling? input) "Whoa, chill out!"
          :else "Whatever.")))
